Avoid dependency cycles in targets that lead to help tags generation.

lyuts opened the pull request:
Fixes issue #3904.

lyuts committed:
Avoid dependency cycles in targets that lead to help tags generation.

jszakmeister commented on runtime/CMakeLists.txt:
```diff
@@ -26,11 +26,15 @@ foreach(DF ${DOCFILES})
+      ${PROJECT_SOURCE_DIR}/runtime/doc ${GENERATED_RUNTIME_DIR}/doc
+    DEPENDS
+    ${DOCFILE_DEPS})
+endforeach()
```
Two things.  First, your repeatedly copying all the files into the directory over and over again--I don't think we want that.

Second, perhaps the better thing to do would be make each docfile depend on it's original source file, instead of this set that gradually dwindles?  We could pitch this loop, and re-use the one above:

```cmake
set(BUILDDOCFILES)
foreach(DF ${DOCFILES})
  get_filename_component(F ${DF} NAME)
  set(BF ${GENERATED_RUNTIME_DIR}/doc/${F})
  list(APPEND BUILDDOCFILES ${BF})
  add_custom_command(OUTPUT ${BF}
    COMMAND ${CMAKE_COMMAND} -E copy ${DF} ${BF}
    DEPENDS ${DF})
endforeach()
```

I think something like this would be better, IMHO, since we're only copying the source file and nothing more for a particular output.

jszakmeister replied:
Note: I didn't try the above code... it may have bugs.

lyuts commented:
@jszakmeister, I agree, copying files all over again is bad. As for dependencies, I wasn't sure how docfile dependencies were determined originally. If you approve following the approach that one doc file depends only on its source file, then I think we should go with this approach as it seems to be much cleaner.

justinmk commented:
> If you approve following the approach that one doc file depends only on its source file,

This should probably be ok, depending on how the tags are generated. I assume the tags are regenerated for the whole tree. If not, then an updated installation could have stale tags, of course.

Let's go with @jszakmeister's approach, but just out of curiosity, does CMake provide a way to treat an entire directory as a single "object" without generating the "cross product" of the contents? Because, that is in fact the only way to model the docs dependency graph without a cycle.

lyuts commented:
@justinkmk I honestly don't know if CMake can do it. I'd have to look it up.

lyuts committed:
Simplify helptags generation targets. Declare dependency in terms of
directory, rather than individual doc files to avoid target dependency
cycles.  This still maintains install targets at doc file level.

lyuts commented:
Taking a second look and considering @justinmk 's suggestion, it looks like we change helptags generation as following (see the latest diff).
Target at line 29 now has no dependencies and probably shouldn't have any. It just copies doc files to a build directory once.
Target at line 34 no longer does copying of doc files but instead depends on target at line 29. This still ensures that before we actually proceed to helptags generation, we do copy them to a build directory.
And lastly, which remains unchanged, target at line 61, i.e. install_helper, handles installation of actual doc files on the system on a doc file level.

Let me know if I'm missing anything with approach. Testing this on my system right now, build and actual package installation, and it seems to work.

justinmk commented:
Merged after squash/commit message fix. Thanks @lyuts . 

Note that this change causes _all_ `/usr/local/share/nvim/runtime/doc/*.txt` artifacts to always re-install. But that's kind of the choice we have to make (alternative is to risk "stale" tags). Previous to this change the up-to-date check did not re-install those files.
